### ISSUE TITLE
Credentials from EKS service account

### ISSUE THREAD

**Issue by sindrig** [aws/aws-sdk-js-v3#1808]
**Describe the bug**
AWS_WEB_IDENTITY_TOKEN_FILE is not utilized in EKS

**SDK version number**
3.0.0

**Is the issue in the browser/Node.js/ReactNative?**
Node.js

**Details of the browser/Node.js/ReactNative version**
bash-5.0# node -v
v12.20.0

**To Reproduce (observed behavior)**
Inside a pod in EKS with AWS_WEB_IDENTITY_TOKEN_FILE defined as a file with a token for the service account
```js
const REGION = 'eu-west-1'
const { STSClient, GetCallerIdentityCommand } = require('@aws-sdk/client-sts')

const sts = new STSClient({ region: REGION })
sts.send(new GetCallerIdentityCommand({})).then(console.log)
```

**Expected behavior**
The service account role should be returned. However, the default cluster role is returned (from the ec2 metadata)

**Additional context**
https://github.com/aws/aws-sdk-js/blob/54f8555bd94d33a1754a44a35286f1d9e31c28a3/lib/credentials/token_file_web_identity_credentials.js#L59

**Comment by AllanZhengYP** [comment]
Hi @sindrig,

Thank you for reporting this. This credential provider is not supported in v3 yet. I have added this one to our high-priority  board. We will post update here once we take it up.

**Comment by kbbqiu** [comment]
This is a blocker for us as well. Looking forward to this release.

**Comment by sgran** [comment]
I've just stumbled across this as well.  Please prioritise the fix.  If we can add a PR to help speed things, we can do so.

**Comment by ingoziemann** [comment]
Is there any workaround for this, otherwise it means 2 weeks of work wasted? 

Do I have overread the warning that the v3 SDK is not production ready?

**Comment by davidhu2000** [comment]
i just ran into this issue as well. v2 sdk works just fine. We are debating if we should downgrade to v2 for the new app we are building. Which isn't ideal given how different v2 and v3 is. Would be good to get some sort of timeline for this issue.

**Comment by nini** [comment]
Is there any workaround possible for this right now? (Except of downgrading to v2 SDK)

**Comment by davidhu2000** [comment]
> Is there any workaround possible for this right now? (Except of downgrading to v2 SDK)

one way we've found is create an IAM user with the env variables

```
export AWS_ACCESS_KEY_ID='EXAMPLE_ID'
export AWS_SECRET_ACCESS_KEY='EXAMPLE_KEY'
```

**Comment by lowjoel** [comment]
@AllanZhengYP while this is being implemented could we please put a note in https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts-minimum-sdk.html please? I spent quite a bit of time debugging why this was not working when the v3 SDK is greater than the required minimum (and would expect what's supported in v2 is supported in v3.)

**Comment by educalleja** [comment]
@trivikr Curious to know if this is something your team is actively working on. We are being blocked by this feature, and are unsure if we should downgrade to v2, which seems to be supported.

**Comment by trivikr** [comment]
The TokenFileWebIdentityCredentials provider was added in v2 in https://github.com/aws/aws-sdk-js/pull/2737
It's missing in JS SDK v3, as called out in our [blog post](https://aws.amazon.com/blogs/developer/modular-aws-sdk-for-javascript-is-now-generally-available/).

> Curious to know if this is something your team is actively working on.

Yes, I'm working on providing this support since this issue has many likes, comments and is blocking multiple customers.

> If so, is there any estimated release date?

We're aiming to release TokenFileWebIdentityCredentials with `3.6.0` (expected release date on 2/18)

**Comment by trivikr** [comment]
> We're aiming to release TokenFileWebIdentityCredentials with `3.6.0` (expected release date on 2/18)

Update: The support for DocumentClient in https://github.com/aws/aws-sdk-js-v3/issues/1223 got prioritized over this issue, as it's blocking more customers.

The work for TokenFileWebIdentityCredentials will be picked up after DocumentClient is delivered.

**Comment by ingoziemann** [comment]
I agree that this Document Client is blocking more users to start migrating, but this issue blocks users that already using the new SDK and workarounds have a big security impact. This issue here you are only discovering if you're done with coding and try to rollout the stuff. For us, that means that we have now to backport the stuff now. So meaning at least two weeks for nothing.  Possible we give the new SDK a new trial in 2 years until its production-ready.

**Comment by carlosescura** [comment]
Is there any update on expected release date for this @trivikr ??
We need to make a decision on whether we should wait for a release or completely refactor our code and use V2 library.

Regards, and thanks for the big effort on building this new library 💪

**Comment by Legion2** [comment]
We already moved back to V2 because this blocked our release.

**Comment by trivikr** [comment]
> Is there any update on expected release date for this ??

The document client will be released this week, and we'll prioritize this request next week.
The release date would be provided once this request is prioritized.

**Comment by trivikr** [comment]
In AWS SDK for JavaScript (v2):
* the token file web identity credentials was added in https://github.com/aws/aws-sdk-js/pull/2737
* It's called after ProcessCredentials and before EC2MetadataCredentials https://github.com/aws/aws-sdk-js/blob/a8e023df608f2c054b33f6fcede06482812e2159/lib/node_loader.js#L68

In v3, the equivalent provider should be called after `fromProcess` in credential-provider-node providerChain https://github.com/aws/aws-sdk-js-v3/blob/3fd14d5e43bc020cdc90932f1ee114f6ffa4ebc1/packages/credential-provider-node/src/index.ts#L53

**Comment by OzoTek** [comment]
I also had this issue. But I couldn't easily downgrade to v2 because of a common dependency with another project that requires the SDK v3.

I implemented the provider myself and it seems to be working fine!
Obviously I'd rather not do that but that's better than having to downgrade for me.

```typescript
import * as fs from 'fs';
import { promisify } from 'util';
import {
  STSClient,
  AssumeRoleWithWebIdentityCommand,
} from '@aws-sdk/client-sts';
import type { Credentials } from '@aws-sdk/types';

const readAsync = promisify(fs.readFile);

const client = new STSClient({
  region: process.env.AWS_REGION || process.env.DEFAULT_REGION || 'us-east-1',
});

export async function webIdentityTokenProvider(): Promise<Credentials> {
  const tokenFilePath = process.env.AWS_WEB_IDENTITY_TOKEN_FILE;
  if (!tokenFilePath) {
    throw new Error(
      'AWS_WEB_IDENTITY_TOKEN_FILE must be provided.',
    );
  }
  const token = await readAsync(tokenFilePath, { encoding: 'ascii' });
  const res = await client.send(
    new AssumeRoleWithWebIdentityCommand({
      RoleArn: process.env.AWS_ROLE_ARN,
      WebIdentityToken: token,
      RoleSessionName: 'assumed-role-session-name',
    }),
  );

  const { Credentials } = res;

  if (!Credentials?.AccessKeyId || !Credentials?.SecretAccessKey) {
    throw new Error('Credentials could not be retrieved.');
  }

  return {
    accessKeyId: Credentials.AccessKeyId,
    secretAccessKey: Credentials.SecretAccessKey,
    expiration: Credentials.Expiration,
    sessionToken: Credentials.SessionToken,
  };
}
```

Usage:

```typescript
// Or any other client for that matter
new DynamoDBClient({
  credentials: webIdentityTokenProvider
});
```

As a note, I had to implement memoization otherwise it'd fetch credentials every time:
```typescript
export function memoizeProvider(
  provider: CredentialProvider,
): CredentialProvider {
  let cache: Credentials | undefined;

  return async function credentialsProvider() {
    if (cache && !areCredentialsExpired(cache)) {
      return cache;
    }

    cache = await provider();
    return cache;
  };
}

function areCredentialsExpired(credentials: Credentials) {
  return (
    (credentials.expiration?.getTime() || Number.POSITIVE_INFINITY) <
    new Date().getTime()
  );
}
```

**Comment by trivikr** [comment]
Update: We're going to create a new package `credential-provider-web-identity` with a function `fromTokenFile` which would implement [TokenFileWebIdentityCredentials](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/TokenFileWebIdentityCredentials.html)

This solution won't work automatically in defaultProvider till https://github.com/aws/aws-sdk-js-v3/issues/1193#issuecomment-795009554 is fixed, but you'll be able to import and call `fromTokenFile` to get credentials.

**Comment by trivikr** [comment]
PR to add credential provider for EKS service account is posted at https://github.com/aws/aws-sdk-js-v3/pull/2147

A separate issue is created for calling credential provider by default in Node.js credential provider chain at https://github.com/aws/aws-sdk-js-v3/issues/2148

**Comment by davidhu2000** [comment]
@trivikr thanks for getting this done. I see the PR merged a few hours ago, when do we expect this to be fully releases for use?

**Comment by trivikr** [comment]
> when do we expect this to be fully releases for use?

The PR https://github.com/aws/aws-sdk-js-v3/pull/2177 will be published with v3.10.0 on Friday 3/26.
We're likely going to push support to read token file from shared config in https://github.com/aws/aws-sdk-js-v3/pull/2178 also in the same release.

Ability to call fromTokenFile automatically is a stretch goal for tomorrow, and it's being tracked in https://github.com/aws/aws-sdk-js-v3/issues/2148

**Comment by davidhu2000** [comment]
Awesome. thank you so much!
